presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayJoin.java

```diff
@@ -0,0 +1,191 @@
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return DESCRIPTION;
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        return specializeArrayJoin(types, arity, typeManager, functionRegistry, ImmutableList.of(false, false));
+    }
+
+    private FunctionInfo specializeArrayJoin(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry, List<Boolean> nullableArguments)
+    {
+        Type type = types.get("T");
+        TypeSignature valueType = type.getTypeSignature();
+        FunctionInfo castFunction = functionRegistry.getExactFunction(internalOperator(CAST.name(), VARCHAR_TYPE_SIGNATURE, ImmutableList.of(valueType)));
+
+        if (!(type instanceof UnknownType)) {
+            checkCondition(castFunction != null, INVALID_FUNCTION_ARGUMENT, "Input type %s not supported", type);
+        }
+
+        TypeSignature arrayType = parameterizedTypeName(StandardTypes.ARRAY, valueType);
+        Signature signature = new Signature(FUNCTION_NAME, VARCHAR_TYPE_SIGNATURE, arrayType, VARCHAR_TYPE_SIGNATURE, VARCHAR_TYPE_SIGNATURE);
+        MethodHandle methodHandle = methodHandle(ArrayJoin.class, "arrayJoin", FunctionInfo.class, Type.class, ConnectorSession.class, Slice.class, Slice.class, Slice.class);
+
+        return new FunctionInfo(signature, getDescription(), isHidden(), methodHandle.bindTo(castFunction).bindTo(type), isDeterministic(), false, nullableArguments);
+    }
+
+    public static Slice arrayJoin(FunctionInfo castFunction, Type elementType, ConnectorSession session, Slice array, Slice delimiter, Slice nullReplacement) throws Throwable
+    {
+        Block arrayBlock = readStructuralBlock(array);
+        int numElements = arrayBlock.getPositionCount();
+        DynamicSliceOutput sliceOutput = new DynamicSliceOutput(array.length());
+        Class<?> javaType = elementType.getJavaType();
+
+        for (int i = 0; i < numElements; i++) {
+            if (arrayBlock.isNull(i)) {
+                sliceOutput.appendBytes(nullReplacement);
+            }
+            else {
+                if (javaType == boolean.class) {
+                    sliceOutput.appendBytes((Slice) castFunction.getMethodHandle().invokeWithArguments(elementType.getBoolean(arrayBlock, i)));
+                }
+                else if (javaType == double.class) {
+                    sliceOutput.appendBytes((Slice) castFunction.getMethodHandle().invokeWithArguments(elementType.getDouble(arrayBlock, i)));
+                }
+                else if (javaType == long.class) {
+                    long value = elementType.getLong(arrayBlock, i);
+                    Class<?>[] parameters = castFunction.getMethodHandle().type().parameterArray();
+                    Slice slice;
+                    if (parameters[0] == ConnectorSession.class) {
```

Comment:
add this ConnectorSession check for the other native container types as well. I don't think we have any others that depend on it right now, but it would be legitimate to have one that uses `Slice.class` or even `boolean` or `double`